flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -505,17 +496,12 @@ private void allocateSlot(TaskManagerSlotInformation taskManagerSlot, JobID jobI
 			resourceManagerId,
 			taskManagerRequestTimeout);
 
-		requestFuture.whenComplete(
+		CompletableFuture<Void> slotAllocationResponseProcessingFuture = requestFuture.handleAsync(
 			(Acknowledge acknowledge, Throwable throwable) -> {
-				if (acknowledge != null) {
-					completableFuture.complete(acknowledge);
-				} else {
-					completableFuture.completeExceptionally(throwable);
+				if (!pendingSlotAllocations.contains(slotId)) {
+					LOG.debug("Ignoring slot allocation update from task executor {} for slot {} and job {}, because the allocation was already completed or cancelled.", instanceId, slotId, jobId);
+					return null;
 				}
-			});
-
-		CompletableFuture<Void> slotAllocationResponseProcessingFuture = completableFuture.handleAsync(
-			(Acknowledge acknowledge, Throwable throwable) -> {
 				if (acknowledge != null) {
 					LOG.trace("Completed allocation of slot {} for job {}.", slotId, jobId);
 					slotTracker.notifyAllocationComplete(slotId, jobId);
```

Comment:
Do we still have to distinguish between `CancellationException` and others in the else branch?

Reply:
hmmmm.....no, it should no longer be possible for the future to be canceled.